File: day03/solution.c
#include <stdio.h>
#include <stdbool.h>

#define NVALUES 1000
#define VALLEN 12

int search(char values[NVALUES][VALLEN+1], bool most)
{
    bool sel[NVALUES];
    
    for(int i = 0; i < NVALUES; i++)
        sel[i] = true;

    for(int i = 0; i <= VALLEN; i++)
    {
        int nsel = 0;

        for(int j = 0; j < NVALUES; j++)
            if(sel[j])
                nsel++;

        if (nsel == 1)
        {
            for(int j = 0; j < NVALUES; j++)
                if(sel[j])
                    return j;
        }
        int n1 = 0;

        for(int j = 0; j < NVALUES; j++)
        {
            if(sel[j] && values[j][i] == '1')
                n1++;
        }

        for(int j = 0; j < NVALUES; j++)
        {
            char tgt = most ? (2*n1 >= nsel ? '1' : '0')
                            : (2*n1 < nsel ? '1' : '0');
            sel[j] = sel[j] && values[j][i] == tgt;
        }
    }

    return -1;
}

int value_to_bin(char val[VALLEN+1])
{
    int n = 0;
    for (int i = 0; i < VALLEN; i++)
        n = 2 * n + (val[i] == '1' ? 1 : 0);
    return n;
}

int main(void)
{
    FILE *fp = fopen("input.txt", "r");
    char values[NVALUES][VALLEN+1];
    for (int i = 0; i < NVALUES; i++)
        fscanf(fp, "%s\n", values[i]);
   
    int most = 0;
    for (int i = 0; i < VALLEN; i++)
    {
        int c1 = 0;
        for(int j = 0; j < NVALUES; j++)
            if (values[j][i] == '1')
                c1++;
        most = 2 * most + (c1>NVALUES/2 ? 1 : 0);
    }
    int least = (1 << 12) - most - 1;
    printf("Part 1 : %d\n", most * least);

    int oxygen = value_to_bin(values[search(values, true)]);
    int co2 = value_to_bin(values[search(values, false)]);
    printf("Part 2 : %d\n", oxygen * co2);
}

File: day03/solution.py
values = [ l.strip() for l in open('input.txt') ]
n = len(values)

b = { True: '1', False : '0' }
mosts = ''.join( b[[ v[i] for v in values ].count('1') > n//2]
        for i in range(12))
most = eval('0b'+mosts)
least = 2**12 - most - 1
print('Part 1', most * least)

def search(val, i, most=True):
    if len(val) == 1:
        return eval('0b'+val[0])

    val1 = [ v for v in val if v[i] == '1' ]
    val0 = [ v for v in val if v[i] == '0' ]
    n1, n0 = len(val1), len(val0)
    if not most:
        val1, val0 = val0, val1

    if n1 >= n0:
        return search(val1, i+1, most=most)
    else:
        return search(val0, i+1, most=most)

oxygen = search(values, 0, True)
co2 = search(values, 0, False)
print('Part 2', oxygen * co2)
